crates/network/src/any/mod.rs

```diff
@@ -46,11 +59,515 @@ impl From<TxType> for AnyTxType {
     }
 }
 
+/// Memoization for deserialization of [`AnyTxEnvelope`] and [`AnyTypedTransaction`].
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[allow(unnameable_types)]
+pub struct DeserMemo {
```

Comment:
I think instead of memoization we could just add `AnyTx` type which would just deserialize all of those if present?

Reply:
i feel like lazy deserialization is more appropriate here, as we are making absolutely no assumptions about the contents of the tx other than that it is some sort of JSON

Reply:
we are making assumptions on both fields name and types they deserialize into in `Transaction` trait implementation, and overhead of always trying to deserialize it wouldn't be that much

imo users of `AnyNetwork` are very unlikely to use it for networks with transaction types which are significantly different from Etheruem transactions and thus require a different deserialization methods. instead, they mostly don't care about the consensus semantics and just want to capture common transaction fields (from,to,data,gas) which are preset for all transactions. and providing easy access to those fields by always trying deserialize them into struct fields would be handy imo

I've took similar approach when embedding header recently in #1573 as it wouldn't make sense to force users deal with raw serde_json values just to get header fields of networks which are missing `mixHash` in it

Reply:
> we are making assumptions on both fields names and types they deserialize into in Transaction trait implementation, and overhead of always trying to deserialize it wouldn't be that much

not exactly. The API in this PR makes no assumptions that the key exists or that its contents are what we expect. If we eagerly deserialize, we have to make some form of those assumptions. Lazy deserialization means we can continue to not assume that, and to allow users to (e.g.) modify the format of an authorization list because they're on the older EIP-7702 specification. If we deserialize eagerly, we end up either consuming that key (and the user is no longer allowed to modify that format at all) or going through a ton of complexity implementing `Deserialize` to preserve the user's access to that key. We want to avoid the case where we render some transaction contents inaccessible due to eager deser attempts

> I've took similar approach when embedding header recently in https://github.com/alloy-rs/alloy/pull/1573 as it wouldn't make sense to force users deal with raw serde_json values just to get header fields of networks which are missing mixHash in it

the approach here with memoization is the same as your approach, except that deserialization is done lazily instead of eagerly. To avoid having users mess around with `serde_json::Value` would it make sense to you to add a convenience function like this:

```
fn deser_key<T>(key: &str) -> Option<T> { ... }
```

Reply:
> the approach here with memoization is the same as your approach, except that deserialization is done lazily instead of eagerly. To avoid having users mess around with `serde_json::Value` would it make sense to you to add a convenience function like this:
> 
> ```
> fn deser_key<T>(key: &str) -> Option<T> { ... }
> ```

I would be fine with that as long as we still have common transaction fields present. imo it would make sense to attempt deserialization of fields of transactions up to 1559 as those are well defined, commonly used and unlikely to change semantics

Reply:
we have deserialization fo fields up to 1559 as `AnyTxEnvelope` accepts the Ethereum tx format. If users are using a custom tx type, we cover access to common fields via the transaction trait

Reply:
`deser_by_key` added in 51eb19a

Reply:
I don't think we have a strong reason to force users to import a trait just to access those fields